### PR TITLE
allow to select key to import instead of using Download folder

### LINKED ISSUE #2658: [core 1.123.0] Import keys from the file rather than directory
Core 1.123.0 will allow using secret key `.asc` file as `param1` path for `DC_IMEX_IMPORT_SELF_KEYS`: https://github.com/deltachat/deltachat-core-rust/pull/4737

Directory selection dialog should be replaced with a file selection dialog.

This is going to solve the problem reported on the forum: https://support.delta.chat/t/deltachat-fails-to-import-keys/2592

### PR DESCRIPTION
**Pull request by adbenitez**
close #2658

### CODE REVIEW

File: src/org/thoughtcrime/securesms/preferences/AdvancedPreferenceFragment.java [hunk 1]
```diff
@@ -330,11 +371,15 @@ private void exportKeys() {
                           .show();
                     }
                     else {
+                        String explain = getActivity().getString(R.string.pref_managekeys_import_explain_tips);
+                        if (Build.VERSION.SDK_INT < 30) { // in older Android we always import from Download folder
+                          explain = getActivity().getString(R.string.pref_managekeys_import_explain_path, DcHelper.getImexDir().getAbsolutePath()) + "\n\n" + explain;
+                        }
                       new AlertDialog.Builder(getActivity())
```
Comment (r10s) [inline src/org/thoughtcrime/securesms/preferences/AdvancedPreferenceFragment.java:378]: sth. is wrong with indentation here - `String explain ...` and `new AlertDialog ...` should have the same indentation

File: src/org/thoughtcrime/securesms/preferences/ListSummaryPreferenceFragment.java [hunk 2]
```diff
@@ -100,10 +100,17 @@ protected void initializeListSummary(ListPreference pref) {
 
   protected ProgressDialog progressDialog = null;
   protected int            progressWhat = 0;
-  protected String         imexDir = "";
+  protected String         pathName = ""; // path/filename displayed to user on finish
   protected boolean        imexUserAborted = false;
```
Comment (r10s) [inline src/org/thoughtcrime/securesms/preferences/ListSummaryPreferenceFragment.java:104]: - these 4 variables are helpers for the second `startImex()` variation and should be moved down (so before `protected void startImex(int what, String imexPath, String pathName)`)

- i find the comment confusing. better drop the comment and improve the variable name to sth. as `pathAsDisplayedToUser` - then things are also clear when the variable is accessed
```

File: src/org/thoughtcrime/securesms/mms/AttachmentManager.java [hunk 3]
```diff
@@ -421,6 +420,24 @@ public boolean isAttachmentPresent() {
     return deck;
   }
 
+  public static @Nullable String getFileName(Context context, Uri uri) {
```
Comment (r10s) [inline src/org/thoughtcrime/securesms/mms/AttachmentManager.java:423]: thanks for diving into that mess 🤿

**Review by r10s — DISMISSED** [review]
some comments

**Comment by r10s** [comment]
wait: i am just checking desktop:

why no just do the same as there:

- open file selector first, if supported, then ask "Import secret keys from "foobar.asc"?\n\n• Existing secret keys will not be deleted ..."  
(if file selector is not supported do the same as before and ask "Import secret keys from "/path/to/download/folder"?\n\n• Existing secret keys will not be deleted ..."  

- after import say "Secret keys imported from "foobar.asc"

the tips do not really explain that a file selector follows, so it feels more natural afterwards.

also, that way, no new strings are needed

**Comment by adbenitez** [comment]
> why no just do the same as there

done :heavy_check_mark:

**Review by r10s — APPROVED** [review]

**Comment by monperrus** [comment]
Hi all, I'm trying to import a public key of a contact, but cannot find the feature, is that possible? thanks!

**Comment by r10s** [comment]
Delta Chat uses [Autocrypt](https://docs.autocrypt.org) and does not want the users be think about keys at all :) - therefore, no, this is not possible, however, this was considered here and then and  of course, this is a topic sometimes.

in general, for feature requests and such discussions, better use the [support forum](https://support.delta.chat)

**Comment by monperrus** [comment]
thanks @r10s 

the feature request is https://support.delta.chat/t/import-and-use-of-existing-gpg-keys/3072/2

importing public GPG keys of contacts is necessary in order to be interoperable with protonmail and gmx/mailevlope.

**Comment by iequidoo** [comment]
> the feature request is https://support.delta.chat/t/import-and-use-of-existing-gpg-keys/3072/2
> 
> importing public GPG keys of contacts is necessary in order to be interoperable with protonmail and gmx/mailevlope.

Btw, i even created a task for this previously: https://github.com/deltachat/deltachat-core-rust/issues/5698. Probably it should be closed and better discussed on the forum first, but the fact is that Delta Chat already imports keys from attachments in incoming emails, so technically it's very doable, just allow the user to trigger the same logic by hand. Though anyway it's a new functionality that should be maintained and covered with tests.

**Comment by monperrus** [comment]
> technically it's very doable, just allow the user to trigger the same logic by hand

great, that would be so good for encryption interoperability. thanks!

**Pull request merged**